packages/openfermion/quri_parts/openfermion/ansatz/kupccgsd.py

```diff
@@ -0,0 +1,109 @@
+from quri_parts.openfermion.utils.add_parametric_pauli_rotation import (
+    add_parametric_pauli_rotation_gate,
+)
+
+
+class KUpCCGSD(ImmutableLinearMappedUnboundParametricQuantumCircuit):
+    """k-unitary pair coupled luster generalized singles and doubles
+    (k-UpCCGSD) ansatz. The ansatz consists of the exponentials of generalized
+    single excitation and pair double excitation operators decomposed by first-
+    order Trotter product formula repeated k times. The term "generalized"
+    means excitation operators created regardless of the occupied orbitals.
+    Here the double excitations are limited to 'pair' excitation, which means
+    two electrons from one spatial orbital to another.
+
+    Ref:
+        Lee et al., Generalized Unitary Coupled Cluster Wave functions for Quantum
+        Computation, J. Chem. Theory Comput. 15, 311–324 (2019),
+        `arXiv:1810.02327 <https://arxiv.org/abs/1810.02327>`_.
+
+        PennyLane's documentations,
+        `qml.kUpCCGSD <https://docs.pennylane.ai/en/stable/
+        code/api/pennylane.kUpCCGSD.html>`_
+
+    Args:
+        n_spin_orbitals: Number of spin orbitals.
+        n_fermions: Number of fermions.
+        k: Number of repetitions.
+        fermion_qubit_mapping: Mapping from :class:`FermionOperator` to
+          :class:`Operator`
+        trotter_number: Number for first-order Trotter product formula.
+        delta_sz: Changes of spin in the excitation.
+    """
+
+    def __init__(
+        self,
+        n_spin_orbitals: int,
+        n_fermions: int,
+        k: int = 1,
+        fermion_qubit_mapping: OpenFermionQubitMapping = jordan_wigner,
+        trotter_number: int = 1,
+        delta_sz: int = 0,
+    ):
+        s_excs = _generalized_single_excitations(n_spin_orbitals, delta_sz)
+        d_excs = _generalized_pair_double_excitations(n_spin_orbitals)
+
+        n_qubits = fermion_qubit_mapping.n_qubits_required(n_spin_orbitals)
+        circuit = LinearMappedUnboundParametricQuantumCircuit(n_qubits)
+
+        op_mapper = fermion_qubit_mapping.get_of_operator_mapper(
+            n_spin_orbitals, n_fermions
+        )
+
+        for _ in range(k):
+            s_exc_params = [
+                circuit.add_parameter(f"theta_s_{i}") for i in range(len(s_excs))
+            ]
+            d_exc_params = [
+                circuit.add_parameter(f"theta_d_{i}") for i in range(len(d_excs))
+            ]
+            for _ in range(trotter_number):
+                add_parametric_pauli_rotation_gate(
+                    circuit, d_excs, d_exc_params, op_mapper, trotter_number
+                )
+                add_parametric_pauli_rotation_gate(
+                    circuit, s_excs, s_exc_params, op_mapper, trotter_number
+                )
+        super().__init__(circuit)
+
+
+def _generalized_single_excitations(
+    n_spin_orbitals: int, delta_sz: int
+) -> Sequence[SingleExcitation]:
+    sz = [0.5 if (i % 2 == 0) else -0.5 for i in range(n_spin_orbitals)]
```

Comment:
Equivalent? (but less readable?)

```suggestion
    sz = [0.5 - (i % 2) for i in range(n_spin_orbitals)]
```